### PR TITLE
add_entity tweaks

### PR DESCRIPTION
**Pull request by RobotCaleb**
lots of goodies here

- grew a new component
- component gen got a tiny bit smarter
- entities without placeable component can't be placed (we have none, currently)
- placeable component dictates what rules placement follows regarding grid snap and rotation amount
- libconfig support for getting and setting generated enums (setting is untested, but seems right)
- fixes rotation to snap when switching between entities with different values
- moved entity switching to cycle_mode
- single press alt_use to step rotate
- long press alt_use to continuously rotate -- This didn't change
- changed rotate direction

### CODE REVIEW

File: gen/gen_enums.py [hunk 1]
```diff
@@ -35,6 +35,8 @@ def main():
         with open(source_file, "w") as source:
             header.write("#pragma once\n")
             header.write("\n")
+            header.write("#include <libconfig.h>\n")
```
Comment (chrisforbes) [inline gen/gen_enums.py]: Need to include the shim after this.

File: src/component/component_managers.h [hunk 2]
```diff
@@ -12,6 +12,7 @@
 #include "gas_producer_component.h"
 #include "light_component.h"
 #include "physics_component.h"
+#include "placeable_component.h"
```
Comment (chrisforbes) [inline src/component/component_managers.h]: Placeable didn't get added until later..

File: gen/gen_enums.py (hunk 1 above)
Comment (RobotCaleb) [reply gen/gen_enums.py]: done

File: src/component/component_managers.h (hunk 2 above)
Comment (RobotCaleb) [reply src/component/component_managers.h]: fixed, thanks

File: src/tools/add_entity.cc [hunk 3]
```diff
@@ -93,47 +101,28 @@ struct add_entity_tool : tool {
...
+            }
+        } while (!entity_stubs[entity_names[entity_name_index]].get_component<placeable_component_stub>());
+
+        auto rotate = get_rotate();
+        if (rotate == 0) {
+            cur_rotate = 0;
+        }
+        else {
+            cur_rotate += rotate / 2;
+            cur_rotate = cur_rotate - (cur_rotate % rotate);
```
Comment (chrisforbes) [inline src/tools/add_entity.cc]: Is this equivalent to the two lines: `cur_rotate -= (cur_rotate + rotate / 2) % rotate` ?

File: src/tools/add_entity.cc [hunk 4]
```diff
@@ -278,15 +209,12 @@ struct add_entity_tool : tool {
                 m[3] = p;
                 break;
             }
-            case PlaceMode::BlockSnapped: {
+            case placement::full_block_snapped: {
                 m = mat_block_face(rc->p, index ^ 1);
```
Comment (chrisforbes) [inline src/tools/add_entity.cc]: Isn't this the same as the step=1 case?

File: src/tools/add_entity.cc (hunk 3 above)
Comment (RobotCaleb) [reply src/tools/add_entity.cc]: probably

File: src/tools/add_entity.cc [hunk 5]
```diff
@@ -203,63 +170,27 @@ struct add_entity_tool : tool {
...
-            case PlaceMode::QuarterBlockSnapped: {
+            case placement::quarter_block_snapped: {
                 step *= 2;
                 // falls through
             }
-            case PlaceMode::HalfBlockSnapped: {
+            case placement::half_block_snapped: {
                 step *= 2;
                 auto pos = rc->intersection;
                 m = mat_rotate_mesh(pos, rc->n);
```
Comment (chrisforbes) [inline src/tools/add_entity.cc]: This rounding logic below should just round all 3 axes. It's simpler, and it tidies up any small inaccuracy in `rc->intersection`

File: src/tools/add_entity.cc [hunk 6]
```diff
@@ -100,39 +100,44 @@ struct add_entity_tool : tool {
...
         auto rotate = get_rotate();
-        if (rotate == 0) {
-            cur_rotate = 0;
-        }
-        else {
-            cur_rotate += rotate / 2;
-            cur_rotate = cur_rotate - (cur_rotate % rotate);
-        }
+
+        cur_rotate -= rotate;
```
Comment (chrisforbes) [inline src/tools/add_entity.cc]: Keep rotation positive, negate when producing the matrix?

File: src/tools/add_entity.cc [hunk 7]
```diff
@@ -100,39 +100,44 @@ struct add_entity_tool : tool {
...
+            }
+        } while (!entity_stubs[entity_names[entity_name_index]].get_component<placeable_component_stub>());
+
+        auto rotate = get_rotate();
+        if (rotate == 0) {
+            cur_rotate = 0;
+        }
+        else {
+            cur_rotate += rotate / 2;
+            cur_rotate = cur_rotate - (cur_rotate % rotate);
```
Comment (chrisforbes) [inline src/tools/add_entity.cc]: Hidden assumption here about rotate>0 ?

Comment (RobotCaleb) [reply src/tools/add_entity.cc]: yes

File: src/tools/add_entity.cc (hunk 5 above)
Comment (RobotCaleb) [reply src/tools/add_entity.cc]: I'm too tired to think about what you mean

File: src/tools/add_entity.cc (hunk 3 above)
Comment (RobotCaleb) [reply src/tools/add_entity.cc]: uh, no, seems to not

Comment (chrisforbes) [reply src/tools/add_entity.cc]: Assume the idea was to snap to the closest valid rotation in the new regime?

File: src/tools/add_entity.cc (hunk 7 above)
Comment (chrisforbes) [reply src/tools/add_entity.cc]: Rather, about cur_rotate>0. This might be OK.

File: src/tools/add_entity.cc (hunk 5 above)
Comment (chrisforbes) [reply src/tools/add_entity.cc]: You're carefully snapping the two tangent axes but not the normal axis. It should be safe to snap all 3.

File: src/tools/add_entity.cc (hunk 4 above)
Comment (RobotCaleb) [reply src/tools/add_entity.cc]: no, because step 1 will not center on block face

File: src/tools/add_entity.cc (hunk 5 above)
Comment (chrisforbes) [reply src/tools/add_entity.cc]: (Not new in this change; just seems like a nice simplification)

File: src/tools/add_entity.cc (hunk 3 above)
Comment (chrisforbes) [reply src/tools/add_entity.cc]: I see, no, it's not quite the same. This is hard to follow because of cur_rotate appearing /everywhere/ in these expressions.

File: src/tools/add_entity.cc (hunk 4 above)
Comment (chrisforbes) [reply src/tools/add_entity.cc]: (Or, if it's not, should it be?)

File: src/tools/add_entity.cc (hunk 3 above)
Comment (RobotCaleb) [reply src/tools/add_entity.cc]: yes. the problem here is shown when going from a rotated not to multiple of 90 entity to the door, which we want axis aligned. the door would rotate properly, but wouldn't be axis aligned.

File: src/tools/add_entity.cc (hunk 4 above)
Comment (chrisforbes) [reply src/tools/add_entity.cc]: OK, that makes sense.

File: src/tools/add_entity.cc (hunk 5 above)
Comment (RobotCaleb) [reply src/tools/add_entity.cc]: Oh. That seems like it might generate the wrong results? I guess it depends on how far `rc->intersection` can ever be?

File: src/tools/add_entity.cc (hunk 6 above)
Comment (RobotCaleb) [reply src/tools/add_entity.cc]: hmm, yeah, could do

File: src/tools/add_entity.cc (hunk 5 above)
Comment (chrisforbes) [reply src/tools/add_entity.cc]: If it's ever > 1/16 of a block I've failed horribly

Comment (RobotCaleb) [reply src/tools/add_entity.cc]: Is this what you mean?
```
                if (n.x != 0) {
                    p.x = std::round(p.x);
                    p.y = std::round(p.y * step) / step;
                    p.z = std::round(p.z * step) / step;
                } else if (n.y != 0) {
                    p.x = std::round(p.x * step) / step;
                    p.y = std::round(p.y);
                    p.z = std::round(p.z * step) / step;
                } else if (n.z != 0) {
                    p.x = std::round(p.x * step) / step;
                    p.y = std::round(p.y * step) / step;
                    p.z = std::round(p.z);
                }```

Comment (RobotCaleb) [reply src/tools/add_entity.cc]: if so, pushed already.

File: src/tools/add_entity.cc (hunk 6 above)
Comment (RobotCaleb) [reply src/tools/add_entity.cc]: done

File: src/tools/add_entity.cc (hunk 5 above)
Comment (chrisforbes) [reply src/tools/add_entity.cc]: I'm saying this should be fine:
```
p.x = std::round(p.x * step) / step;
p.y = std::round(p.y * step) / step;
p.z = std::round(p.z * step) / step;
```

If we're ever off the surface by more than `step/2` then we've got bigger issues.

Comment (RobotCaleb) [reply src/tools/add_entity.cc]: done

**Comment by RobotCaleb** [comment]
I think all review bits have been covered
